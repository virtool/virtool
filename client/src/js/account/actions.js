/**
 * Redux action creators for use with the active user account.
 *
 * @module account/actions
 */
import { simpleActionCreator } from "../utils";
import {
    GET_ACCOUNT,
    UPDATE_ACCOUNT,
    UPDATE_ACCOUNT_SETTINGS,
    CHANGE_ACCOUNT_PASSWORD,
    GET_API_KEYS,
    CREATE_API_KEY,
    UPDATE_API_KEY,
    REMOVE_API_KEY,
    LOGOUT,
    CLEAR_API_KEY
} from "../actionTypes";

/**
 * Returns action that can trigger an API call for getting the current account data.
 *
 * @func
 * @returns {object}
 */
export const getAccount = simpleActionCreator(GET_ACCOUNT.REQUESTED);

/**
 * Returns an action that can trigger an API call for updating the current account.
 *
 * @func
 * @param update {object} the update to apply to the account
 * @returns {object}
 */
export const updateAccount = (update) => ({
    type: UPDATE_ACCOUNT.REQUESTED,
    update
});

/**
 * Returns an action that can trigger an API call for updating the settings for the current account.
 *
 * @func
 * @param update {object} the update to apply to the account
 * @returns {object}
 */
export const updateAccountSettings = (update) => ({
    type: UPDATE_ACCOUNT_SETTINGS.REQUESTED,
    update
});

/**
 * Returns an action that can trigger an API call for changing the password for the current account.
 *
 * @func
 * @param oldPassword {string} the old account password (used for verification)
 * @param newPassword {string} the new account password
 * @returns {object}
 */
export const changePassword = (oldPassword, newPassword) => ({
    type: CHANGE_ACCOUNT_PASSWORD.REQUESTED,
    oldPassword,
    newPassword
});

/**
 * Returns action that can trigger an API call for getting the API keys owned by the current account.
 *
 * @func
 * @returns {object}
 */
export const getAPIKeys = simpleActionCreator(GET_API_KEYS.REQUESTED);

/**
 * Returns action that can trigger an API call for creating a new API key for the current account.
 *
 * @func
 * @param name {string} a name for the API key
 * @param permissions {object} permissions configuration object for the new API key
 * @returns {object}
 */
export const createAPIKey = (name, permissions) => ({
    type: CREATE_API_KEY.REQUESTED,
    name,
    permissions
});

/**
 * Clears temporarily stored new API keys.
 *
 * @func
 * @returns {object}
 */
export const clearAPIKey = simpleActionCreator(CLEAR_API_KEY);

/**
 * Returns action that can trigger an API call for updating the permissions for an API key owned by the current account.
 *
 * @func
 * @param keyId {string} the unique id for the API key
 * @param permissions {object} permissions configuration object for the new API key
 * @returns {object}
 */
export const updateAPIKey = (keyId, permissions) => ({
    type: UPDATE_API_KEY.REQUESTED,
    keyId,
    permissions
});

/**
 * Returns action that can trigger an API call for removing an API key owned by the current account.
 *
 * @func
 * @param keyId {string} the unique id for the API key
 * @returns {object}
 */
export const removeAPIKey = (keyId) => ({
    type: REMOVE_API_KEY.REQUESTED,
    keyId
});

/**
 * Returns action that can trigger an API call that will logout the current session.
 *
 * @func
 * @returns {object}
 */
export const logout = simpleActionCreator(LOGOUT.REQUESTED);
