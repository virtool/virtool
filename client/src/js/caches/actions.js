import { GET_CACHE } from "../app/actionTypes";

export const getCache = cacheId => ({
    type: GET_CACHE.REQUESTED,
    cacheId
});
