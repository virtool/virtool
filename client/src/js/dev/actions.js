import { POST_DEV_COMMAND } from "../app/actionTypes";

export const postDevCommand = command => ({ type: POST_DEV_COMMAND.REQUESTED, command });
