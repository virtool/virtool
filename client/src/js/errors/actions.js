/**
 * Redux action creator for clearing errors.
 *
 * @module error/actions
 */
import { CLEAR_ERROR } from "../actionTypes";

/**
 * Returns action that clears specific error in the store.
 *
 * @func
 * @returns {object}
 */
export const clearError = (error) => ({
    type: CLEAR_ERROR,
    error
});
