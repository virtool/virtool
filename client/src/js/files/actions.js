/**
 * Redux action creators for use with file management.
 *
 * @module files/actions
 */

import { simpleActionCreator } from "../utils";
import {
    WS_UPDATE_FILE,
    WS_REMOVE_FILE,
    FIND_FILES,
    REMOVE_FILE,
    UPLOAD,
    UPLOAD_PROGRESS,
    HIDE_UPLOAD_OVERLAY
} from "../actionTypes";

/**
 * Returns an action that should be dispatched when a file document is updated via websocket.
 *
 * @func
 * @param data {object} the data passed in the websocket message
 * @returns {object}
 */
export const wsUpdateFile = (data) => ({
    type: WS_UPDATE_FILE,
    data
});

/**
 * Returns an action that should be dispatched when a file document is removed via websocket.
 *
 * @func
 * @param data {object} the data passed in the websocket message
 * @returns {object}
 */
export const wsRemoveFile = (data) => ({
    type: WS_REMOVE_FILE,
    data
});

/**
 * Returns an action that can trigger an API call that finds file documents.
 *
 * @func
 * @param fileType {string} a file type to filter the returned documents by
 * @param page {number} which page of results to return
 * @returns {object}
 */
export const findFiles = (fileType, page) => ({
    type: FIND_FILES.REQUESTED,
    fileType,
    page
});

/**
 * Returns an action that can trigger the upload of a file to the server.
 *
 * @func
 * @param localId {string} the local id for the upload (NOT the fileId)
 * @param file {object} file to be uploaded
 * @param fileType {string} file type
 * @param onProgress {function} callback function for uploadProgress API call
 * @returns {object}
 */
export const upload = (localId, file, fileType, onProgress) => ({
    type: UPLOAD.REQUESTED,
    localId,
    file,
    fileType,
    onProgress
});

/**
 * Returns an action that can trigger an API call that removes a file by its fileId.
 *
 * @func
 * @param fileId {string} the unique id for the file
 * @returns {object}
 */
export const removeFile = (fileId) => ({
    type: REMOVE_FILE.REQUESTED,
    fileId
});

/**
 * Returns and action that updates the progress of an ongoing upload.
 *
 * @func
 * @param localId {string} the local id for the upload (NOT the fileId)
 * @param progress {number} the new progress value
 * @returns {object}
 */
export const uploadProgress = (localId, progress) => ({
    type: UPLOAD_PROGRESS,
    localId,
    progress
});

/**
 * A simple action that causes the upload dialog overlay to be hidden.
 *
 * @func
 * @returns {object}
 */
export const hideUploadOverlay = simpleActionCreator(HIDE_UPLOAD_OVERLAY);
