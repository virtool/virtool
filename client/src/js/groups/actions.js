import { simpleActionCreator } from "../utils";
import { LIST_GROUPS, CREATE_GROUP, SET_GROUP_PERMISSION, REMOVE_GROUP } from "../actionTypes";

/**
 * Returns an action that triggers a request for all groups from the API.
 *
 * @func
 */
export const listGroups = simpleActionCreator(LIST_GROUPS.REQUESTED);

/**
 * Returns an action that triggers a API request to create a group with the given `groupId`.
 *
 * @func
 * @param groupId {string} the id for the new group
 * @returns {object}
 */
export const createGroup = (groupId) => ({
    type: CREATE_GROUP.REQUESTED,
    groupId
});

/**
 * Returns action that can trigger an API request for modifying group permissions.
 *
 * @param groupId {string} the id for the specific group
 * @param permission {string} the specific permission field
 * @param value {boolean} is checked or not
 * @returns {object}
 */
export const setGroupPermission = (groupId, permission, value) => ({
    type: SET_GROUP_PERMISSION.REQUESTED,
    groupId,
    permission,
    value
});

/**
 * Returns an action that triggers a API request to remove a group with the given `groupId`.
 *
 * @func
 * @param groupId {string} the id for the new group
 * @returns {object}
 */
export const removeGroup = (groupId) => ({
    type: REMOVE_GROUP.REQUESTED,
    groupId
});
