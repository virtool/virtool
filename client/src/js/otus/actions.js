import {
    ADD_ISOLATE,
    ADD_SEQUENCE,
    CREATE_OTU,
    EDIT_ISOLATE,
    EDIT_OTU,
    EDIT_SEQUENCE,
    FIND_OTUS,
    GET_OTU,
    GET_OTU_HISTORY,
    HIDE_OTU_MODAL,
    REMOVE_ISOLATE,
    REMOVE_OTU,
    REMOVE_SEQUENCE,
    REVERT,
    SELECT_ISOLATE,
    SET_ISOLATE_AS_DEFAULT,
    SHOW_ADD_ISOLATE,
    SHOW_EDIT_ISOLATE,
    SHOW_EDIT_OTU,
    SHOW_REMOVE_ISOLATE,
    SHOW_REMOVE_OTU,
    SHOW_REMOVE_SEQUENCE,
    WS_INSERT_OTU,
    WS_REMOVE_OTU,
    WS_UPDATE_OTU
} from "../app/actionTypes";
import { simpleActionCreator } from "../utils/utils";

export const wsInsertOTU = data => ({
    type: WS_INSERT_OTU,
    data
});

export const wsUpdateOTU = data => ({
    type: WS_UPDATE_OTU,
    data
});

export const wsRemoveOTU = data => ({
    type: WS_REMOVE_OTU,
    data
});

export const findOTUs = (refId, term, verified, page) => ({
    type: FIND_OTUS.REQUESTED,
    refId,
    term,
    verified,
    page
});

/**
 * Returns action that can trigger an API call for retrieving a specific OTU.
 *
 * @func
 * @param otuId {string} unique OTU id
 * @returns {object}
 */
export const getOTU = otuId => ({
    type: GET_OTU.REQUESTED,
    otuId
});

/**
 * Returns action that can trigger an API call for getting a OTU's history.
 *
 * @func
 * @param otuId {string} unique OTU id
 * @returns {object}
 */
export const getOTUHistory = otuId => ({
    type: GET_OTU_HISTORY.REQUESTED,
    otuId
});

/**
 * Returns action that can trigger an API call for creating a new OTU.
 *
 * @func
 * @param name {string} unique OTU name
 * @param abbreviation {string} unique abbreviation for OTU name
 * @returns {object}
 */
export const createOTU = (refId, name, abbreviation) => ({
    type: CREATE_OTU.REQUESTED,
    refId,
    name,
    abbreviation
});

/**
 * Returns action that can trigger an API call for modifying a OTU.
 *
 * @func
 * @param otuId {string} unique OTU id
 * @param name {string} unique OTU name
 * @param abbreviation {string} unique abbreviation of OTU name
 * @param schema {array} array of sequences in custom order
 * @returns {object}
 */
export const editOTU = (otuId, name, abbreviation, schema) => ({
    type: EDIT_OTU.REQUESTED,
    otuId,
    name,
    abbreviation,
    schema
});

/**
 * Returns action that can trigger an API call for removing a OTU.
 *
 * @func
 * @param otuId {string} unique OTU id
 * @param history {object} list of all changes made to the OTU
 * @returns {object}
 */
export const removeOTU = (refId, otuId, history) => ({
    type: REMOVE_OTU.REQUESTED,
    refId,
    otuId,
    history
});

/**
 * Returns action that can trigger an API call for adding an isolate to a OTU.
 *
 * @func
 * @param otuId {string} unique OTU id
 * @param sourceType {string} category of isolate source types
 * @param sourceName {string} the name of the isolate source
 * @returns {object}
 */
export const addIsolate = (otuId, sourceType, sourceName) => ({
    type: ADD_ISOLATE.REQUESTED,
    otuId,
    sourceType,
    sourceName
});

/**
 * Returns action that can trigger an API call for modifying which isolate is made default.
 *
 * @func
 * @param otuId {string} unique OTU id
 * @param isolateId {string} unique isolate id
 * @returns {object}
 */
export const setIsolateAsDefault = (otuId, isolateId) => ({
    type: SET_ISOLATE_AS_DEFAULT.REQUESTED,
    otuId,
    isolateId
});

/**
 * Returns action that can trigger an API call for modifying an isolate.
 *
 * @func
 * @param otuID {string} unique OTU id
 * @param isolateId {string} unique isolate id
 * @param sourceType {string} category of isolate source types
 * @param sourceName {string} the name of the isolate source
 * @returns {object}
 */
export const editIsolate = (otuId, isolateId, sourceType, sourceName) => ({
    type: EDIT_ISOLATE.REQUESTED,
    otuId,
    isolateId,
    sourceType,
    sourceName
});

/**
 * Returns action that can trigger an API call for removing an isolate from a OTU.
 *
 * @func
 * @param otuId {string} unique OTU id
 * @param isolateId {string} unique isolate id
 * @param nextIsolateId {string} if removed isolate was default,
 * first in resulting list (i.e. the next isolate) becomes default
 * @returns {object}
 */
export const removeIsolate = (otuId, isolateId, nextIsolateId) => ({
    type: REMOVE_ISOLATE.REQUESTED,
    otuId,
    isolateId,
    nextIsolateId
});

/**
 * Returns action that can trigger an API call for adding a sequence to an isolate.
 *
 * @func
 * @param otuId {string} unique OTU id
 * @param isolateId {string} unique isolate id
 * @param accession {string} the accession for the sequence
 * @param definition {string} descriptive definition of the sequence
 * @param host {string} the host the sequence originated from
 * @param sequence {string} an abbreviation for the OTU
 * @param segment {string} the schema segment associated with the OTU
 * @param target {string} the reference target associated with the sequence
 * @returns {object}
 */
export const addSequence = (otuId, isolateId, accession, definition, host, sequence, segment, target) => ({
    type: ADD_SEQUENCE.REQUESTED,
    otuId,
    isolateId,
    accession,
    definition,
    host,
    sequence,
    segment,
    target
});

/**
 * Returns action that can trigger an API call for modifying a sequence.
 *
 * @func
 * @param otuId {string} unique OTU id
 * @param isolateId {string} unique isolate id
 * @param sequenceId {string} unique sequence id
 * @param accession {string} a new accession for the sequence
 * @param definition {string} descriptive definition of the sequence
 * @param host {string} the host the sequence originated from
 * @param sequence {string} an abbreviation for the OTU
 * @param segment {string} the schema segment associated with the OTU
 * @returns {object}
 */
export const editSequence = (otuId, isolateId, sequenceId, accession, definition, host, sequence, segment) => ({
    type: EDIT_SEQUENCE.REQUESTED,
    otuId,
    isolateId,
    sequenceId,
    accession,
    definition,
    host,
    sequence,
    segment
});

/**
 * Returns action that can trigger an API call for removing a sequence from an isolate.
 *
 * @func
 * @param otuId {string} unique OTU id
 * @param isolateId {string} unique isolate id
 * @param sequenceId {string} unique sequence id
 * @returns {object}
 */
export const removeSequence = (otuId, isolateId, sequenceId) => ({
    type: REMOVE_SEQUENCE.REQUESTED,
    otuId,
    isolateId,
    sequenceId
});

/**
 * Returns action that can trigger an API call for deleting unbuilt changes of a OTU.
 *
 * @func
 * @param otuId {string} unique OTU id
 * @param version {string} OTU index version
 * @returns {object}
 */
export const revert = (otuId, otuVersion, changeId) => ({
    type: REVERT.REQUESTED,
    otuId,
    otuVersion,
    change_id: changeId
});

/**
 * Returns action for selecting an isolate to view.
 *
 * @func
 * @param isolateId {string} unique isolate id
 * @returns {object}
 */
export const selectIsolate = isolateId => ({
    type: SELECT_ISOLATE,
    isolateId
});

/**
 * Returns action for displaying the edit OTU modal.
 *
 * @func
 * @returns {object}
 */
export const showEditOTU = simpleActionCreator(SHOW_EDIT_OTU);

/**
 * Returns action for displaying the remove OTU modal.
 *
 * @func
 * @returns {object}
 */
export const showRemoveOTU = simpleActionCreator(SHOW_REMOVE_OTU);

/**
 * Returns action for displaying the add isolate modal.
 *
 * @func
 * @returns {object}
 */
export const showAddIsolate = simpleActionCreator(SHOW_ADD_ISOLATE);

/**
 * Returns action for displaying the edit isolate modal.
 *
 * @func
 * @param otuId {string} unique OTU id
 * @param isolateId {string} unique isolate id
 * @param sourceType {string} category of isolate source types
 * @param sourceName {string} the name for the isolate source
 * @returns {object}
 */
export const showEditIsolate = (otuId, isolateId, sourceType, sourceName) => ({
    type: SHOW_EDIT_ISOLATE,
    otuId,
    isolateId,
    sourceType,
    sourceName
});

/**
 * Returns action for displaying the remove isolate modal.
 *
 * @func
 * @returns {object}
 */
export const showRemoveIsolate = simpleActionCreator(SHOW_REMOVE_ISOLATE);

/**
 * Returns action for displaying the remove sequence modal.
 *
 * @func
 * @param sequenceId {string} unique sequence id
 * @returns {object}
 */
export const showRemoveSequence = sequenceId => ({
    type: SHOW_REMOVE_SEQUENCE,
    sequenceId
});

/**
 * Returns action for hiding the OTU modal.
 *
 * @func
 * @returns {object}
 */
export const hideOTUModal = simpleActionCreator(HIDE_OTU_MODAL);
