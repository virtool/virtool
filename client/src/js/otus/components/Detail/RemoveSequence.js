import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { removeSequence, hideOTUModal } from "../../actions";
import { RemoveModal } from "../../../base";

class RemoveSequence extends React.Component {

    handleConfirm = () => {
        this.props.onConfirm(this.props.otuId, this.props.isolateId, this.props.sequenceId);
    };

    render () {
        const removeMessage = (
            <span>Are you sure you want to remove the sequence
                <strong> {this.props.sequenceId}</strong> from
                <strong> {this.props.isolateName}</strong>?
            </span>
        );

        return (
            <RemoveModal
                name={`${this.props.sequenceId}`}
                noun="Sequence"
                onConfirm={this.handleConfirm}
                onHide={this.props.onHide}
                show={Boolean(this.props.sequenceId)}
                message={removeMessage}
            />
        );
    }
}

RemoveSequence.propTypes = {
    otuId: PropTypes.string,
    isolateId: PropTypes.string,
    isolateName: PropTypes.string,
    sequenceId: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
    onHide: PropTypes.func,
    onConfirm: PropTypes.func
};

const mapStateToProps = state => ({
    sequenceId: state.otus.removeSequence
});

const mapDispatchToProps = dispatch => ({

    onHide: () => {
        dispatch(hideOTUModal());
    },

    onConfirm: (otuId, isolateId, sequenceId) => {
        dispatch(removeSequence(otuId, isolateId, sequenceId));
    }

});

export default connect(mapStateToProps, mapDispatchToProps)(RemoveSequence);
