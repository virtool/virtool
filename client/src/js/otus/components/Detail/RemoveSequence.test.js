import RemoveSequence from "./RemoveSequence";
import * as actions from "../../actions";

describe("<RemoveSequence />", () => {
    const initialState = { otus: { removeSequence: "test-sequence" } };
    const store = mockStore(initialState);
    const props = {
        otuId: "123abc",
        isolateId: "456def",
        isolateName: "test-isolate"
    };
    let wrapper;

    it("renders correctly", () => {
        wrapper = shallow(<RemoveSequence store={store} {...props} />).dive();
        expect(wrapper).toMatchSnapshot();
    });

    describe("dispatch functions", () => {
        let spy;

        beforeAll(() => {
            wrapper = shallow(<RemoveSequence store={store} {...props} />).dive();
        });

        afterEach(() => {
            spy.restore();
        });

        it("Closing modal dispatches hideOTUModal() action", () => {
            spy = sinon.spy(actions, "hideOTUModal");
            expect(spy.called).toBe(false);

            wrapper.prop("onHide")();

            expect(spy.calledOnce).toBe(true);
        });

        it("Clicking Confirm button dispatches removeSequence() action", () => {
            spy = sinon.spy(actions, "removeSequence");
            expect(spy.called).toBe(false);

            wrapper.prop("onConfirm")();

            expect(spy.calledWith(
                props.otuId,
                props.isolateId,
                initialState.otus.removeSequence
            )).toBe(true);
        });

    });

});
