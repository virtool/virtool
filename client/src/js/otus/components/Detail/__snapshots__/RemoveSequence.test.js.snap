// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RemoveSequence /> renders correctly 1`] = `
<RemoveModal
  message={
    <span>
      Are you sure you want to remove the sequence
      <strong>
         
        test-sequence
      </strong>
       from
      <strong>
         
        test-isolate
      </strong>
      ?
    </span>
  }
  name="test-sequence"
  noun="Sequence"
  onConfirm={[Function]}
  onHide={[Function]}
  show={true}
/>
`;
